src/layers/VolumeLayer/VolumeLayer.js

```diff
@@ -0,0 +1,154 @@
+ * @property {string=} renderingMode One of Maximum Intensity Projection, Minimum Intensity Projection, or Additive
+ * @property {Object=} modelMatrix A column major affine transformation to be applied to the volume.
+ * @property {Array.<number>=} xSlice 0-1 interval on which to slice the volume.
+ * @property {Array.<number>=} ySlice 0-1 interval on which to slice the volume.
+ * @property {Array.<number>=} zSlice 0-1 interval on which to slice the volume.
+ * @property {function=} onViewportLoad Function that gets called when the data in the viewport loads.
+ */
+
+/**
+ * @type {{ new(...props: LayerProps[]) }}
+ */
+const VolumeLayer = class extends CompositeLayer {
+  updateState({ changeFlags, oldProps, props }) {
+    const { propsChanged } = changeFlags;
+    const loaderChanged =
+      typeof propsChanged === 'string' && propsChanged.includes('props.loader');
+    const loaderSelectionChanged =
+      props.loaderSelection !== oldProps.loaderSelection;
+    // Only fetch new data to render if loader has changed
+    if (loaderChanged || loaderSelectionChanged) {
+      const {
+        loader,
+        loaderSelection = [],
+        resolution,
+        onViewportLoad
+      } = this.props;
+      const source = loader[resolution];
+      let progress = 0;
+      const totalRequests =
+        // eslint-disable-next-line no-bitwise
+        (source.shape[source.labels.indexOf('z')] >> resolution) *
+        loaderSelection.length;
+      const onUpdate = () => {
+        progress += 0.5 / totalRequests;
+        this.setState({ progress });
+      };
+      const volumePromises = loaderSelection.map(selection =>
+        getVolume({
+          selection,
+          source,
+          onUpdate,
+          downsampleDepth: 2 ** resolution
+        })
+      );
+
+      Promise.all(volumePromises).then(volumes => {
+        if (onViewportLoad) {
+          onViewportLoad(volumes);
+        }
+        const volume = {
+          data: volumes.map(d => d.data),
+          width: volumes[0].width,
+          height: volumes[0].height,
+          depth: volumes[0].depth
+        };
+
+        this.setState({ ...volume });
+      });
+    }
+  }
+
+  renderLayers() {
+    const { loader, id, resolution } = this.props;
+    const { dtype } = loader[resolution];
+    const { data, width, height, depth, progress } = this.state;
+    if (!(width && height)) {
+      const { viewport } = this.context;
+      return new TextLayer({
+        id: `loading-text-layer-${id}`,
+        coordinateSystem: COORDINATE_SYSTEM.CARTESIAN,
+        data: [
+          {
+            text: `Loading Volume ${String((progress || 0) * 100).slice(
+              0,
+              5
+            )}% (Firefox/Chrome Only)...`,
+            position: viewport.position
+          }
+        ],
+        getColor: [220, 220, 220, 255],
+        getSize: 25,
+        sizeUnits: 'meters',
+        sizeScale: 2 ** -viewport.zoom
+      });
+    }
```

Comment:
I am not sure how we would provide default functionality to a layer that involves creating a div outside of DeckGL, so I think a deck.gl layer makes sense for now as the default.  I am open to other suggestions for default behavior (and then how we would configure not using that default behavior, like truthiness of `onLoading`?), but I think otherwise we should return to this question when we get to the Avivator portion of this PR because we may want `onLoading` functions for every channel (i.e for when they are in the process of changing) as well.